### PR TITLE
[quant][fx][improvement] Add lowering support for BatchNormQuantizeHandler

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72490
* #72489

Summary:
This is an effort to move the current implementation towards the reference quantized model design:
https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md
so that we use reference model in the default fbgemm/qnnpack path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps.test_qbatch_norm

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34062365](https://our.internmc.facebook.com/intern/diff/D34062365)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 1]
```diff
@@ -41,13 +41,21 @@ def is_fixed_qparams_node(node, modules):
...
+LOWER_MODULE_MAP: Dict[Type[torch.nn.Module], Type[ReferenceableQuantizedModule]] = {
     nnqr.Linear: nnq.Linear,
     nnqr.Conv1d: nnq.Conv1d,
     nnqr.Conv2d: nnq.Conv2d,
     nnqr.Conv3d: nnq.Conv3d,
 }
 
+# TODO: merge with LOWER_MODULE_MAP after we merge
+# _lower_weighted_ref_module and special_pattern_replacement
+LOWER_MODULE_MAP2 = {
```
Comment (vkuzo) [inline torch/ao/quantization/fx/_lower_to_native_backend.py:54]: maybe name it `LOWER_MODULE_MAP_TEMP`?

File: torch/nn/quantized/modules/batchnorm.py [hunk 2]
```diff
@@ -26,6 +26,19 @@ def from_float(cls, mod):
...
+    def from_reference(cls, bn, output_scale, output_zero_point):
+        qbn = cls(
+            bn.num_features,
+            bn.eps,
+            bn.momentum,
+            device=bn.weight.device,
+            dtype=bn.weight.dtype
+        )
+        qbn.scale = float(output_scale)
+        qbn.zero_point = float(output_zero_point)
```
Comment (vkuzo) [inline torch/nn/quantized/modules/batchnorm.py:39]: should this be int?

**Review by vkuzo — APPROVED** [review]

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 1 above)
Comment (andrewor14) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:54]: or maybe `SPECIAL_PATTERN_LOWER_MODULE_MAP` if we're using it for `special_pattern_replacement` only

**Review by andrewor14 — APPROVED** [review]

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 3]
```diff
@@ -41,13 +41,21 @@ def is_fixed_qparams_node(node, modules):
     return is_call_function, is_call_method, is_call_module
 
 # Mapping from reference module class to the replacement quantized module class for lowering
-LOWER_MODULE_MAP: Dict[Type[nn.Module], Type[ReferenceableQuantizedModule]] = {
+# TODO: fix typing, the key is reference module
+LOWER_MODULE_MAP: Dict[Type[torch.nn.Module], Type[ReferenceableQuantizedModule]] = {
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:45]: I see, yeah I'm not sure why we need this type, I thought it's supposed to be referencemodule, should we remove `ReferenceableQuantizedModule`?

[the remark this answers is not in the document]
Comment (andrewor14) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:45]: Well the problem is we need to call `from_reference` in `_lower_weighted_ref_module`, but a `torch.nn.Module` doesn't have this method, so I feel we do need a common abstract class to declare this method

[the remark this answers is not in the document]
Comment (andrewor14) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:45]: (Either that or we check `hasattr(module, "from_reference")`, but I feel the abstract class is cleaner)

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34062365).
